Bounds inflation unit tests

craigtaverner opened the pull request:
Originally this PR added an alternative algorithm to BoundedGeoHexGridTiler, but performance tests showed that it was not actually better than the original, so now the PR only adds a number of unit tests.

craigtaverner committed:
Alternative algorithm for inflated bounds

The BoundedGeoHexGridTiler inflates the bounding box for parent
cell searches since the parent cells bounds do not match the
descendents.

This algorithm does a slightly more refined calculation of the
inflation factor by considering all four corners and inflating in
the four directions independently.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

iverase reviewed (changes requested):
See my comment, you need to be very careful with bounding boxes crossing the dateline.

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/BoundedGeoHexGridTiler.java:
```diff
@@ -42,52 +42,67 @@ public BoundedGeoHexGridTiler(int resolution, GeoBoundingBox bbox) {
+        final double maxY = Math.min(wAvg(bbox.top(), boundsOfCells.getMaxY(), factor), 90d);
+        final double left = GeoUtils.normalizeLon(wAvg(bbox.left(), boundsOfCells.getMinX(), factor));
+        final double right = GeoUtils.normalizeLon(wAvg(bbox.right(), boundsOfCells.getMaxX(), factor));
+        if (right - left >= 360d) {
```
Yes, I've since written some unit tests, and they do indeed fail. I'll see if the fix is easy enough.

craigtaverner committed:
Remove alternative algorithm for bbox inflation

The new algorithm was faster than the old for the same inflation factor
but had more failures, and a longer tail of failures (which needed
higher inflation factors). So it would take a lot more work (probably)
to find and fix these edge cases. An easier option is to keep the
original algorithm.

craigtaverner committed:
Reduce bbox inflation factor to 0.25

This increases performance by about 10% (a factor of 0.22 increased
performance by about 13%).

We found the error rate for `testGeoShapeGeoHex` was about 0.5% for
inflation factor 0.19 and 0.2% for inflation factor 0.22.
These last 2/1000 failed tests required inflation factors of 0.27
0.30 respectively, so we'll investigate them individually.

craigtaverner committed:
Increase bbox inflation factor to 0.35

Since there were three failures discovered for 0.25, all of which passed
at 0.30, we decided to set the inflation factor to 0.35 initially for
the sake of caution, and consider revising it down again later with
more careful testing.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexGridTiler.java:
```diff
@@ -249,20 +249,22 @@ private static class BoundedGeoHexGridTiler extends GeoHexGridTiler {
+         * This is done by taking the H3 cells at two corners, and expanding the filter width
+         * by 50% of the max width of those cells, and filter height by 50% of the max height of those cells.
+         *
+         * The inflation factor of 50% has been verified using test GeoHexTilerTests#testLargeShapeWithBounds
```
This comment is not correct any longer?

craigtaverner replied:
If you are talking about the last line, I changed that from `tuned` to `verified`, as I thought that was correct. But would you prefer we remove the line entirely?

iverase replied:
I was thinking more that the inflation factor is not 50% any more but 35%?

craigtaverner replied:
Ah, of course. I fixed that now.

craigtaverner committed:
Fixed comment about bounded tiler inflation

craigtaverner committed:
Fixed comment about bounded tiler inflation

iverase reviewed (approved):
LGTM
